dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/compute/functions/ServerWithNatRuleToNodeMetadata.java

```diff
@@ -0,0 +1,123 @@
+import org.jclouds.dimensiondata.cloudcontrol.domain.State;
+import org.jclouds.dimensiondata.cloudcontrol.domain.internal.ServerWithExternalIp;
+import org.jclouds.domain.Credentials;
+import org.jclouds.domain.Location;
+import org.jclouds.domain.LoginCredentials;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Strings.nullToEmpty;
+import static com.google.common.collect.Iterables.find;
+import static org.jclouds.location.predicates.LocationPredicates.idEquals;
+
+@Singleton
+public class ServerWithNatRuleToNodeMetadata implements Function<ServerWithExternalIp, NodeMetadata> {
+
+   public static final Map<State, NodeMetadata.Status> serverStateToNodeStatus = ImmutableMap.<State, NodeMetadata.Status>builder()
+         .put(State.PENDING_DELETE, NodeMetadata.Status.PENDING).put(State.PENDING_CHANGE, NodeMetadata.Status.PENDING)
+         .put(State.FAILED_ADD, NodeMetadata.Status.ERROR).put(State.FAILED_CHANGE, NodeMetadata.Status.ERROR)
+         .put(State.FAILED_DELETE, NodeMetadata.Status.ERROR).put(State.DELETED, NodeMetadata.Status.TERMINATED)
+         .put(State.NORMAL, NodeMetadata.Status.RUNNING).put(State.UNRECOGNIZED, NodeMetadata.Status.UNRECOGNIZED)
+         .build();
+
+   private final Supplier<Set<? extends Location>> locations;
+   private final GroupNamingConvention nodeNamingConvention;
+   private final BaseImageToImage baseImageToImage;
+   private final BaseImageToHardware baseImageToHardware;
+   private final Map<String, Credentials> credentialStore;
+   private final DimensionDataCloudControlApi api;
+
+   @Inject
+   ServerWithNatRuleToNodeMetadata(@Memoized final Supplier<Set<? extends Location>> locations,
+         final GroupNamingConvention.Factory namingConvention, final BaseImageToImage baseImageToImage,
+         final BaseImageToHardware baseImageToHardware, final Map<String, Credentials> credentialStore,
+         final DimensionDataCloudControlApi api) {
+      this.nodeNamingConvention = checkNotNull(namingConvention, "namingConvention").createWithoutPrefix();
+      this.locations = checkNotNull(locations, "locations");
+      this.baseImageToImage = checkNotNull(baseImageToImage, "baseImageToImage");
+      this.baseImageToHardware = checkNotNull(baseImageToHardware, "osImageToHardware");
+      this.credentialStore = checkNotNull(credentialStore, "credentialStore cannot be null");
+      this.api = checkNotNull(api, "api cannot be null");
+   }
+
+   @Override
+   public NodeMetadata apply(final ServerWithExternalIp serverWithExternalIp) {
+      NodeMetadataBuilder builder = new NodeMetadataBuilder();
+      Server server = serverWithExternalIp.server();
+      builder.ids(server.id());
+      builder.name(server.name());
+      builder.hostname(serverWithExternalIp.server().description());
+      if (server.datacenterId() != null) {
```

Comment:
Under which circumstances can this be null?

Reply:
Should not be null. Our xsd has it as a required attribute:
`<xs:attribute name="datacenterId" type="xs:string" use="required"/>`